Contributing.md

```diff
@@ -27,6 +28,42 @@ To work and develop PyO3, you need Python & Rust installed on your system.
 
 * When using pyenv on macOS, installing a Python version using `--enable-shared` is required to make it work. i.e `env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.7.12`
 
+### Testing, linting, etc. with nox
+
+[`Nox`][nox] is used to automate many of our CI tasks and can be used locally to handle verfication tasks as you code. We recommend running these actions via nox to make use of our prefered configuration options. You can install nox into your global python with pip: `pip install nox` or (recommended) with [`pipx`][pipx] `pip install pipx`, `pipx install nox`
+
+The main nox commands we have implemented are:
+
+* `nox -s test` will run the full suite of recommended rust and python tests (>10 minutes)
+* `nox -s test-rust -- skip-full` will run a short suite of rust tests (2-3 minutes)
+* `nox -s ruff` will check python linting and apply standard formatting rules
+* `nox -s rustfmt` will check basic rust linting and apply standard formatting rules
+* `nox -s clippy` will run clippy to make recommendations on rust style
+* `nox -s bench` will benchmark your rust code
+* `nox -s codspeed` will run our suite of rust and python performance tests
+* `nox -s coverage` will analyse test coverage and output `coverage.json` (alternatively: `nox -s coverage lcov` outputs `lcov.info`)
+* `nox -s check-guide` will use [`lychee`][lychee] to check all the links in the guide and doc comments.
+
+Use  `nox -l` to list the full set of subcommands you can run.
+
+#### UI Tests
+
+PyO3 uses [`trybuild`][trybuild] to develop UI tests to capture error messages from the Rust compiler for some of the macro functionality.
+
+Because there are several feature combinations for these UI tests, when updating them all (e.g. for a new Rust compiler version) it may be helpful to use the `update-ui-tests` nox session:
+
+```bash
+nox -s update-ui-tests
+```
+
+### Github workflow
+
+In addition to the larger CI workflow which runs on each Pull Request, we provide a short CI workflow which runs on every push to github. This workflow runs the nox sessions: `rustfmt`, `ruff`, `clippy` and `test-rust -- skip-full` and is designed to run in under 200s to provide feedback within a reasonable timeframe and not use all your github actions minutes.
+
+In order to save resources this workflow is opt-in. If you wish to make use of the workflow then please name your branch `ci/...`.
```

Comment:
@adamreichold, @Icxolu - Thank you for the clarification, now I understand! Would an opt-in approach be an acceptable way to allow those whos personal workflow prefers github without forcing this on anyone else?

Reply:
As written, I personally am _not_ in favour of pushing this into forks via the main PyO3 repository, whether opt-in or opt-out. Having this in the main fork provides a lot of persuasion to use such a centralized workflow, c.f. the "we provide a short CI workflow" wording, and I do not want that.

Especially since we do provide an alternative on which I would prefer people to focus on, i.e. the Nox sessions. Finally, a commit adding that CI pipeline can be carried in forks indefinitely if desired.

Reply:
I'm not sure I understand your comment "Finally, a commit adding that CI pipeline can be carried in forks indefinitely if desired."

The problem with not including this in the main repo is that you are making it very difficult for forks to effectively add their own ci if they wish. Any workflow which a fork adds for its own use is included in their PRs - and so needs removing manually every time.

Personally, I don't want to run 2-5 manual commands and wait minutes on my local machine before each push. I push many times in one hour and this would make development very painful.

Reply:
> Any workflow which a fork adds for its own use is included in their PRs - and so needs removing manually every time.

I think compared to the effort expected for reviewing one's own changes before pushing them for others to review, dropping a single commit is very little work.  

> Personally, I don't want to run 2-5 manual commands and wait minutes on my local machine before each push. I push many times in one hour and this would make development very painful.

Then let's improve the Nox sessions so it becomes a single command if that is the underlying issue.

Reply:
Ok. Thanks for engaging so constructively, it helps both my understanding and the code base.

I think that nox command should balance
- running enough checks to catch common oversights (I'm assuming that people have already tried to check the errors they could imagine happening)
- running quickly so as to not delay development flow 

I personally notice a psychological barrier at about 200s (3 mins plus a bit of "wiggle"). What is your feeling on this? 

In terms of the checks to cover, you have much more experience than I do with the code base. Add a new contributor, the "silly issues" I was most thankful to have checked were:

- `fmt`
- `check-guide`

I also felt a lot more confident knowing that: 

- `clippy`
- `test-rust -- skip-full`

were guaranteed to run.

Reply:
Personally I rely heavily on IDE feedback while writing, so `rustfmt` and `clippy` are run pretty often automatically. Once I finished my implementation I usually just run `nox -x` to run the default sessions with "fail early" config and fix any errors that pop up. 

Maybe you can propose in a followup what sessions work best for your workflow (your list above seems pretty reasonable to me, maybe `ruff` could be added since it's also fairly quick) and then we can see whether we want to add a new session for these quick feedback actions or if we want to change the default group and introduce a new session for the full/long run, more intended to run before submitting the PR. What do you think about that?